pkg/kubelet/device-plugin/manager.go

```diff
@@ -0,0 +1,178 @@
+	"golang.org/x/net/context"
+	"google.golang.org/grpc"
+
+	pluginapi "k8s.io/kubernetes/pkg/kubelet/apis/device-plugin/v1alpha1"
+)
+
+// NewManager creates a new manager on the socket `socketPath` and can
+// rebuild state from devices and available []Device.
+// f is the callback that is called when a device becomes unhealthy
+// socketPath is present for testing purposes in production this is pluginapi.KubeletSocket
+func NewManagerImpl(socketPath string, f MonitorCallback) (*ManagerImpl, error) {
+	glog.V(2).Infof("Creating Device Plugin manager at %s", socketPath)
+
+	if socketPath == "" || !filepath.IsAbs(socketPath) {
+		return nil, fmt.Errorf(ErrBadSocket+" %v", socketPath)
+	}
+
+	dir, file := filepath.Split(socketPath)
+	return &ManagerImpl{
+		Endpoints: make(map[string]*endpoint),
+
+		socketname: file,
+		socketdir:  dir,
+		callback:   f,
+	}, nil
+}
+
+func (m *ManagerImpl) Start() error {
+	glog.V(2).Infof("Starting Device Plugin manager")
+
+	socketPath := filepath.Join(m.socketdir, m.socketname)
+	os.MkdirAll(m.socketdir, 0755)
+
+	if err := os.Remove(socketPath); err != nil && !os.IsNotExist(err) {
+		glog.Errorf(ErrRemoveSocket+" %+v", err)
+		return err
+	}
+
+	s, err := net.Listen("unix", socketPath)
+	if err != nil {
+		glog.Errorf(ErrListenSocket+" %+v", err)
+		return err
+	}
+
+	m.server = grpc.NewServer([]grpc.ServerOption{}...)
+
+	pluginapi.RegisterRegistrationServer(m.server, m)
+	go m.server.Serve(s)
+
+	return nil
+}
+
+// Devices is the map of devices that are known by the Device
+// Plugin manager with the Kind of the devices as key
+func (m *ManagerImpl) Devices() map[string][]*pluginapi.Device {
+	glog.V(2).Infof("Devices called")
+
+	m.mutex.Lock()
+	defer m.mutex.Unlock()
+
+	devs := make(map[string][]*pluginapi.Device)
+	for k, e := range m.Endpoints {
+		glog.V(2).Infof("Endpoint: %+v: %+v", k, e)
+		e.mutex.Lock()
+		devs[k] = copyDevices(e.devices)
+		e.mutex.Unlock()
+	}
+
+	return devs
+}
+
+// Allocate is the call that you can use to allocate a set of Devices
+func (m *ManagerImpl) Allocate(resourceName string,
+	devs []*pluginapi.Device) (*pluginapi.AllocateResponse, error) {
+
+	m.mutex.Lock()
+	defer m.mutex.Unlock()
+
+	if len(devs) == 0 {
+		return nil, nil
+	}
+
+	glog.Infof("Recieved request for devices %v for device plugin %s",
+		devs, resourceName)
+
+	e, ok := m.Endpoints[resourceName]
+	if !ok {
+		return nil, fmt.Errorf("Unknown Device Plugin %s", resourceName)
+	}
+
+	return e.allocate(devs)
+}
+
+func (m *ManagerImpl) Register(ctx context.Context,
+	r *pluginapi.RegisterRequest) (*pluginapi.Empty, error) {
+
+	glog.V(2).Infof("Got request for Device Plugin %s", r.ResourceName)
+
+	if r.Version != pluginapi.Version {
+		return &pluginapi.Empty{},
+			fmt.Errorf(pluginapi.ErrUnsuportedVersion)
+	}
+
+	if err := IsResourceNameValid(r.ResourceName); err != nil {
+		return &pluginapi.Empty{}, err
+	}
+
+	// TODO check for conflicting labels
+
+	if _, ok := m.Endpoints[r.ResourceName]; ok {
```

Comment:
Say a device plugin restarts, would this checking prevent the new device plugin from registering with the running kubelet?